geps/gep-1767/index.md

```diff
@@ -0,0 +1,701 @@
+    //
+    // A wildcard indicates that the requests with all HTTP methods are allowed.
+    // When responding to a credentialed requests, the gateway must specify 
+    // one or more HTTP methods in the value of the Access-Control-Allow-Methods response header, 
+    // instead of specifying the * wildcard.
+    //
+    // Config:
+    //   allowMethods: ["*"]
+    // 
+    // Output:
+    //   Access-Control-Allow-Methods: ["*"]
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +listType=set
+    // +kubebuilder:validation:MaxItems=16
+    AllowMethods []string `json:"allowMethods,omitempty"`
+ 
+    // AllowHeaders indicates which HTTP request headers are supported 
+    // for accessing the requested resource.
+    // Header names are not case sensitive.
+    //
+    // Config:
+    //   allowHeaders: ["DNT", "Keep-Alive", "User-Agent", "X-Requested-With", "If-Modified-Since", "Cache-Control", "Content-Type", "Range", "Authorization"]
+    // Output:
+    //   Access-Control-Allow-Headers: DNT, Keep-Alive, User-Agent, X-Requested-With, If-Modified-Since, Cache-Control, Content-Type, Range, Authorization
+    //
+    // A wildcard indicates that the requests with all HTTP headers are allowed.
+    // When responding to a credentialed requests, the gateway must specify 
+    // one or more HTTP headers in the value of the Access-Control-Allow-Headers response header, 
+    // instead of specifying the * wildcard.
+    //
+    // Config:
+    //   allowHeaders: ["*"]
+    // Output:
+    //   Access-Control-Allow-Headers: *
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +listType=set
+    // +kubebuilder:validation:MaxItems=64
+    AllowHeaders []string `json:"allowHeaders,omitempty"`
+
+    // ExposeHeaders indicates which HTTP response headers are exposed to clients 
+    // for the cross-origin request that is not a "preflight" request.
+    // Header names are not case sensitive.
+    //
+    // Config:
+    //   exposeHeaders: ["Content-Security-Policy"]
+    // Output:
+    //   Access-Control-Expose-Headers: Content-Security-Policy
+    //
+    // A wildcard indicates that the responses with all HTTP headers are exposed to clients.
+    // When responding to a credentialed requests, the gateway must specify 
+    // one or more HTTP headers in the value of the Access-Control-Expose-Headers response header, 
+    // instead of specifying the * wildcard.
+    //
+    // Config:
+    //   exposeHeaders: ["*"]
+    // Output:
+    //   Access-Control-Expose-Headers: *
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +listType=set
+    // +kubebuilder:validation:MaxItems=64
+    ExposeHeaders []string `json:"exposeHeaders,omitempty"`
+
+    // MaxAge indicates the duration (in seconds) for the client to cache 
+    // the results of a "preflight" request. 
+    //
+    // The default value of header Access-Control-Max-Age is 5 (seconds).
```

Comment:
Based on [Fetch Living Standard](https://fetch.spec.whatwg.org), the default value of Access-Control-Max-Age is 5 (seconds).

If a HTTPCORSFilter does NOT set  'MaxAge' specifically, the gateway will return the header "Access-Control-Max-Age: 5" to clients by default.

For example, the default value of Access-Control-Max-Age for [Ingress-NGINX `Enable CORS`](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#enable-cors) is 1728000 seconds.

Reply:
The question I have:  if unset, should the server _respond_ `Access-Control-Max-Age: 5`, or should the server not include anything and let the client decide (where it will likely default to 5s)?

Envoy prior art is "not include"

Reply:
Since the default value for [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age) is 5 seconds, I think it's good to set an explicit default here: in case the gateway does not include such a header, it's up to the client such a decision, which is never a good thing in my opinion. Setting a default here makes the whole thing explicit and always defined.